text/0000-target-stages.md

```diff
@@ -0,0 +1,146 @@
+
+# Summary
+[summary]: #summary
+
+Communication between compiler activities via the incremental system to avoid more recompilations than necessary.
+
+# Motivation
+[motivation]: #motivation
+
+The current model for incremental recompilations doesn't share progress between compiler activities, leading to unnecessary rebuilds. Users notice redundant compilations, as
+"Changes in workspaces trigger unnecessary rebuilds" was submitted as [a big complaint in compiler performance][perf-survey].
+
+Introducing a concept of truly incremental compilation helps for these cases which share a common base and would help with:
+
+- `target` directory size; artifacts are shared between activities with a common base (e.g. `check` and `build` on the same crate and same flags)
+- Compilation times, as e.g. `check` -> `build` allows building to start right from type-checking, and allows `check` to be skipped altogether.
+
+The proposed solution is to add a notion of "compiler target stages", which would be served into `rustc` usually via Cargo with a `--target-stage=<ast|macro-expansion|hir|analyze-hir|...>`
+flag. This flag would be denoted a special role in the compiler and it would have a special place in the incremental system.
+
+The flag denotes which target the compiler can go up to, and serves as an early indicator of how the build was performed. This moves the burden of stage-indication from the compiler flags, and gives a reliant way for future compilations to evaluate if they should run, load a dependency graph into memory, or start from scratch altogether.
+
+Disconnecting compiler flags from an early recompilation judgement allows us to almost completely overlap the various
+compiler activities of different levels, and allow for more flexibility in activities other than `build` and `check`,
+such as `clippy` or other existing and future projects that may need to partially build with the Rust compiler.
+
+---
+
+While not all cases overlap with "lower" compiler-adjacent activites (`check` would be lower than `build`) and those
+cases would need a recompilation anyway, this allows us to [perform further optimizations][#further-optimizations] than just holding the two workflows. 
+
+# Guide-level explanation
+[guide-level-explanation]: #guide-level-explanation
+
+
+The `--target-stage` flag is an option that instructs the compiler to only go so far when compiling your code. This
+allows the compiler to perform numerous optimizations to avoid redundant recompilations.
+
+Users don't usually have to worry about this flag, as it's handled automatically by Cargo via `cargo check` and `cargo build`.
+For an explanation for RFC readers, see [Motivation](#motivation).
+
+# Reference-level explanation
+[reference-level-explanation]: #reference-level-explanation
+
+There are two parts to this feature, one handled in the compiler itself and another handled by the incremental system
+and heavily interacting with Cargo.
```

Comment:
I'm a bit unclear on what Cargo will do with this new compiler feature.

Reply:
Edited that portion, should I add a dedicated section about this?

Reply:
So this is about improving the tracking of what inputs invalidates the incremental cache?  How does callers being able to specify arbitrary target stages and `cargo check` specifying a specific one fit into that?

Reply:
> Yes, **instead of Cargo** handling it with compiler flags, the more complex parts would be handled by this new flag.

The part in emphasis is new to me.  What is Cargo no longer supposed to do?  Why does the RFC not talk about this?

> These two concepts go in the same RFC because they both modify the incremental system in pretty major ways, so splitting them would be unhelpful and inefficient.

From reading the RFC, it was not clear to me that the proposal is for two different features that happen to both touch incremental compilation.